File: internal/db/models/jail/jail.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package jailModels

import (
	"fmt"
	"time"

	networkModels "github.com/alchemillahq/sylve/internal/db/models/network"
	"github.com/alchemillahq/sylve/pkg/utils"
	"gorm.io/gorm"
)

func (Network) TableName() string {
	return "jail_networks"
}

type Network struct {
	ID     uint `gorm:"primaryKey" json:"id"`
	JailID uint `json:"jid" gorm:"column:jid;index"`

	Name string `json:"name" gorm:"not null;uniqueIndex:idx_jail_network_name"`

	SwitchID   uint   `json:"switchId" gorm:"not null;index"`
	SwitchType string `json:"switchType" gorm:"index;not null;default:standard"`

	StandardSwitch *networkModels.StandardSwitch `gorm:"-" json:"standardSwitch,omitempty"`
	ManualSwitch   *networkModels.ManualSwitch   `gorm:"-" json:"manualSwitch,omitempty"`

	MacID         *uint                 `json:"macId" gorm:"column:mac_id"`
	MacAddressObj *networkModels.Object `json:"macObj" gorm:"foreignKey:MacID"`

	IPv4ID    *uint                 `json:"ipv4Id" gorm:"column:ipv4_id"`
	IPv4Obj   *networkModels.Object `json:"ipv4Obj" gorm:"foreignKey:IPv4ID"`
	IPv4GwID  *uint                 `json:"ipv4GwId" gorm:"column:ipv4_gw_id"`
	IPv4GwObj *networkModels.Object `json:"ipv4GwObj" gorm:"foreignKey:IPv4GwID"`

	IPv6ID    *uint                 `json:"ipv6Id" gorm:"column:ipv6_id"`
	IPv6Obj   *networkModels.Object `json:"ipv6Obj" gorm:"foreignKey:IPv6ID"`
	IPv6GwID  *uint                 `json:"ipv6GwId" gorm:"column:ipv6_gw_id"`
	IPv6GwObj *networkModels.Object `json:"ipv6GwObj" gorm:"foreignKey:IPv6GwID"`

	DefaultGateway bool `json:"defaultGateway" gorm:"default:false"`

	DHCP  bool `json:"dhcp" gorm:"default:false"`
	SLAAC bool `json:"slaac" gorm:"default:false"`
}

func (n *Network) AfterFind(tx *gorm.DB) error {
	switch n.SwitchType {
	case "standard":
		var s networkModels.StandardSwitch
		if err := tx.
			Preload("Ports").
			Preload("AddressObj").
			Preload("AddressObj.Entries").
			Preload("AddressObj.Resolutions").
			Preload("Address6Obj").
			Preload("Address6Obj.Entries").
			Preload("Address6Obj.Resolutions").
			Preload("NetworkObj").
			Preload("NetworkObj.Entries").
			Preload("NetworkObj.Resolutions").
			Preload("Network6Obj").
			Preload("Network6Obj.Entries").
			Preload("Network6Obj.Resolutions").
			Preload("GatewayAddressObj").
			Preload("GatewayAddressObj.Entries").
			Preload("GatewayAddressObj.Resolutions").
			Preload("Gateway6AddressObj").
			Preload("Gateway6AddressObj.Entries").
			Preload("Gateway6AddressObj.Resolutions").
			First(&s, n.SwitchID).Error; err != nil {
			return fmt.Errorf("load standard switch %d: %w", n.SwitchID, err)
		}
		n.StandardSwitch = &s

	case "manual":
		var m networkModels.ManualSwitch
		if err := tx.First(&m, n.SwitchID).Error; err != nil {
			return fmt.Errorf("load manual switch %d: %w", n.SwitchID, err)
		}
		n.ManualSwitch = &m

	default:
		return fmt.Errorf("unknown switch type: %s", n.SwitchType)
	}

	return tx.Preload("Entries").
		Where("id IN ?", []uint{
			utils.GetVal(n.MacID),
			utils.GetVal(n.IPv4ID),
			utils.GetVal(n.IPv4GwID),
			utils.GetVal(n.IPv6ID),
			utils.GetVal(n.IPv6GwID),
		}).
		Find(&[]networkModels.Object{}).
		Error
}

type JailStats struct {
	ID          uint    `json:"id" gorm:"primaryKey"`
	JailID      uint    `json:"jid" gorm:"column:jid;index"`
	CPUUsage    float64 `json:"cpuUsage"`
	MemoryUsage float64 `json:"memoryUsage"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
}

func (j JailStats) GetID() uint {
	return j.ID
}

func (j JailStats) GetCreatedAt() time.Time {
	return j.CreatedAt
}

type Storage struct {
	ID     uint   `gorm:"primaryKey" json:"id"`
	JailID uint   `json:"jid" gorm:"column:jid;index"`
	Pool   string `json:"pool" gorm:"not null"`
	GUID   string `json:"guid" gorm:"uniqueIndex;not null"`
	Name   string `json:"name"`
	IsBase bool   `json:"isBase" gorm:"default:false"`
}

func (Storage) TableName() string {
	return "jail_storages"
}

type JailHookPhase string

const (
	JailHookPhasePreStart  JailHookPhase = "prestart"
	JailHookPhaseStart     JailHookPhase = "start"
	JailHookPhasePostStart JailHookPhase = "poststart"
	JailHookPhasePreStop   JailHookPhase = "prestop"
	JailHookPhaseStop      JailHookPhase = "stop"
	JailHookPhasePostStop  JailHookPhase = "poststop"
)

type JailHooks struct {
	ID      uint          `json:"id" gorm:"primaryKey"`
	JailID  uint          `json:"jid" gorm:"column:jid;index"`
	Phase   JailHookPhase `json:"phase" gorm:"column:phase;index"`
	Enabled bool          `json:"enabled"`
	Script  string        `json:"script"`
}

type JailSnapshot struct {
	ID uint `json:"id" gorm:"primaryKey"`

	JailID uint `json:"jid" gorm:"column:jid;index;uniqueIndex:idx_jail_snapshot_unique,priority:1"`
	CTID   uint `json:"ctId" gorm:"column:ct_id;index"`

	ParentSnapshotID *uint `json:"parentSnapshotId" gorm:"column:parent_snapshot_id;index"`

	Name        string `json:"name" gorm:"not null"`
	Description string `json:"description" gorm:"default:''"`

	SnapshotName string `json:"snapshotName" gorm:"column:snapshot_name;not null;uniqueIndex:idx_jail_snapshot_unique,priority:2"`
	RootDataset  string `json:"rootDataset" gorm:"column:root_dataset;not null"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}

func (JailSnapshot) TableName() string {
	return "jail_snapshots"
}

type JailType string

const (
	JailTypeFreeBSD JailType = "freebsd"
	JailTypeLinux   JailType = "linux"
)

type Jail struct {
	ID          uint     `json:"id" gorm:"primaryKey"`
	CTID        uint     `json:"ctId" gorm:"unique;not null;uniqueIndex"`
	Name        string   `json:"name" gorm:"not null;unique"`
	Hostname    string   `json:"hostname"`
	Description string   `json:"description"`
	Type        JailType `json:"type"`

	StartAtBoot *bool `json:"startAtBoot" gorm:"default:false"`
	StartOrder  int   `json:"startOrder"`

	InheritIPv4 bool `json:"inheritIPv4"`
	InheritIPv6 bool `json:"inheritIPv6"`

	ResourceLimits *bool  `json:"resourceLimits" gorm:"default:true"`
	Cores          int    `json:"cores"`
	CPUSet         []int  `json:"cpuSet" gorm:"serializer:json;type:json"`
	Memory         int    `json:"memory"`
	DevFSRuleset   string `json:"devfsRuleset"`

	Fstab             string      `json:"fstab"`
	CleanEnvironment  bool        `json:"cleanEnvironment"`
	AdditionalOptions string      `json:"additionalOptions"`
	AllowedOptions    []string    `json:"allowedOptions" gorm:"serializer:json;type:json"`
	JailHooks         []JailHooks `json:"jailHooks" gorm:"foreignKey:JailID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`

	Storages  []Storage      `json:"storages" gorm:"foreignKey:JailID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT"`
	Networks  []Network      `json:"networks" gorm:"foreignKey:JailID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT"`
	Snapshots []JailSnapshot `json:"snapshots,omitempty" gorm:"foreignKey:JailID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
	Stats     []JailStats    `json:"-" gorm:"foreignKey:JailID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`

	MetadataMeta string `json:"metadataMeta"`
	MetadataEnv  string `json:"metadataEnv"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`

	StartLogs string     `json:"startLogs" gorm:"default:''"`
	StopLogs  string     `json:"stopLogs" gorm:"default:''"`
	StartedAt *time.Time `json:"startedAt" gorm:"default:null"`
	StoppedAt *time.Time `json:"stoppedAt" gorm:"default:null"`
}

File: web/src/lib/types/jail/jail.ts
import { z } from 'zod/v4';
import { NetworkObjectSchema } from '../network/object';

export interface CreateData {
    name: string;
    hostname: string;
    id: number;
    node: string;
    description: string;
    storage: {
        pool: string;
        base: string;
        fstab: string;
    };
    network: {
        switch: string;
        mac: number;
        inheritIPv4: boolean;
        inheritIPv6: boolean;
        ipv4: number;
        ipv4Gateway: number;
        ipv6: number;
        ipv6Gateway: number;
        dhcp: boolean;
        slaac: boolean;
    };
    hardware: {
        cpuCores: number;
        ram: number;
        startAtBoot: boolean;
        resourceLimits: boolean;
        bootOrder: number;
        devfsRuleset: string;
    };
    advanced: {
        jailType: 'linux' | 'freebsd';
        additionalOptions: string;
        cleanEnvironment: boolean;
        execScripts: Record<ExecPhaseKey, ExecPhaseState>;
        allowedOptions: string[];
        metadata: {
            env: string;
            meta: string;
        };
    };
}
export const JailStorageSchema = z.object({
    id: z.number().int(),
    jid: z.number().int(),
    pool: z.string(),
    guid: z.string(),
    name: z.string(),
    isBase: z.boolean()
});
export const SimpleJailSchema = z.object({
    id: z.number().int(),
    name: z.string(),
    ctId: z.number().int(),
    state: z.enum(['ACTIVE', 'INACTIVE', 'UNKNOWN']).optional()
});

export const NetworkSchema = z.object({
    id: z.number().int(),
    jid: z.number().int(),
    name: z.string(),
    switchId: z.number().int(),
    switchType: z.enum(['standard', 'manual']),
    macId: z.number().int().nullable(),
    macObj: NetworkObjectSchema.optional().nullable(),
    ipv4Id: z.number().int().nullable(),
    ipv4GwId: z.number().int().nullable(),
    ipv6Id: z.number().int().nullable(),
    ipv6GwId: z.number().int().nullable(),
    dhcp: z.boolean().nullable().default(false),
    slaac: z.boolean().nullable().default(false),
    defaultGateway: z.boolean().default(false)
});

export const JailHookPhaseSchema = z.enum([
    'prestart',
    'start',
    'poststart',
    'prestop',
    'stop',
    'poststop'
]);

export const JailHookSchema = z.object({
    phase: JailHookPhaseSchema,
    enabled: z.boolean(),
    script: z.string()
});

const JailBaseSchema = SimpleJailSchema.extend({
    description: z.string().nullable(),
    startAtBoot: z.boolean(),
    startOrder: z.number().int(),
    inheritIPv4: z.boolean(),
    inheritIPv6: z.boolean(),
    networks: z.array(NetworkSchema).optional().default([]),
    storages: z.array(JailStorageSchema).optional().default([]),
    type: z.enum(['freebsd', 'linux']),
    fstab: z.string(),
    devfsRuleset: z.string(),
    additionalOptions: z.string(),
    allowedOptions: z.array(z.string()).default([]),
    hooks: z
        .array(JailHookSchema)
        .nullable()
        .optional()
        .transform((value) => value ?? []),
    metadataMeta: z.string(),
    metadataEnv: z.string()
});

export const JailSchema = z.preprocess((input) => {
    if (input && typeof input === 'object' && !Array.isArray(input)) {
        const payload = { ...(input as Record<string, unknown>) };
        if (payload.hooks == null) {
            if (Array.isArray(payload.JailHooks)) {
                payload.hooks = payload.JailHooks;
            } else if (Array.isArray(payload.jailHooks)) {
                payload.hooks = payload.jailHooks;
            }
        }
        return payload;
    }

    return input;
}, JailBaseSchema);

export const JailStateSchema = z.object({
    ctId: z.number().int(),
    state: z.enum(['ACTIVE', 'INACTIVE', 'UNKNOWN']),
    pcpu: z.number(),
    memory: z.number()
});

export const JailLogsSchema = z.object({
    logs: z.string()
});

export const JailStatSchema = z.object({
    id: z.number().int(),
    jid: z.number().int(),
    cpuUsage: z.number(),
    memoryUsage: z.number(),
    createdAt: z.string()
});

export const ExecPhaseDefs = [
    {
        key: 'prestart',
        label: 'Pre-start (exec.prestart)',
        description: 'Runs on the host before the jail starts'
    },
    {
        key: 'start',
        label: 'Start (exec.start)',
        description: 'Runs inside the jail to start it, Often /bin/sh /etc/rc'
    },
    {
        key: 'poststart',
        label: 'Post-start (exec.poststart)',
        description: 'Runs inside the jail after it has started'
    },
    {
        key: 'prestop',
        label: 'Pre-stop (exec.prestop)',
        description: 'Runs inside the jail before it is stopped'
    },
    {
        key: 'stop',
        label: 'Stop (exec.stop)',
        description: 'Runs inside the jail to stop it, Often /bin/sh /etc/rc.shutdown'
    },
    {
        key: 'poststop',
        label: 'Post-stop (exec.poststop)',
        description: 'Runs on the host after the jail has stopped'
    }
] as const;

export type ExecPhaseKey = (typeof ExecPhaseDefs)[number]['key'];
export interface ExecPhaseState {
    enabled: boolean;
    script: string;
}

export type SimpleJail = z.infer<typeof SimpleJailSchema>;
export type Jail = z.infer<typeof JailSchema>;
export type JailStorage = z.infer<typeof JailStorageSchema>;
export type JailNetwork = z.infer<typeof NetworkSchema>;
export type JailHook = z.infer<typeof JailHookSchema>;
export type JailState = z.infer<typeof JailStateSchema>;
export type JailLogs = z.infer<typeof JailLogsSchema>;
export type JailStat = z.infer<typeof JailStatSchema>;
